Sources/Bot/Services/MergeService.swift

```diff
@@ -402,27 +402,36 @@ extension MergeService {
                     .observe(on: scheduler)
             case .blocked,
                  .unstable:
-                return github.fetchCommitStatus(for: metadata.reference)
-                    .flatMap(.latest) { commitStatus -> SignalProducer<Event, AnyError> in
-                        switch commitStatus.state {
-                        case .pending:
-                            return .value(.integrationDidChangeStatus(.updating, metadata))
-                        case .failure:
-                            return .value(.integrationDidChangeStatus(.failed(.checksFailing), metadata))
-                        case  .success:
-                            return github.fetchPullRequest(number: metadata.reference.number)
-                                .map { metadata in
-                                    switch metadata.mergeState {
-                                    case .clean:
-                                        return .retryIntegration(metadata)
-                                    default:
-                                        return .integrationDidChangeStatus(.failed(.blocked), metadata)
+                let pullRequest = metadata.reference
+                return github.fetchAllStatusChecks(for: pullRequest).map { statusChecks -> Bool in
+                    return statusChecks.map({ $0.state }).contains(.pending)
+                }.flatMap(.latest) { pendingStatusChecks -> SignalProducer<Event, AnyError> in
+                    if pendingStatusChecks {
+                        return .value(.integrationDidChangeStatus(.updating, metadata))
+                    } else {
+                        return github.fetchCommitStatus(for: metadata.reference)
```

Comment:
Not completely sure here (still need my morning coffee 😄) but couldn't we derive the `switch` below from the `statusChecks` that we fetched in `fetchAllStatusChecks?
I mean, instead of doing 2 API calls (fetching all status checks, then if none pending, fetch commit status and switch on that), maybe there's a way to only do the `fetchAllStatusChecks` API call and determine everything from there?

Maybe the `static func combinedState(for states: [CommitState.State]) -> CommitState.State` method that we already have in `CommitState.swift` can even help in that?

Reply:
I wouldn't go there considering that what we are using is undocumented APIs and inferring will meant that we will be trying to replicate undocumented logic. I also wanted to make as little changes as possible, so kept the rest of logic the same.

Reply:
I don't think we can avoid two API calls in this case but we need to re-work this logic a bit because for the failure we need to check if we want all checks to pass or just the required checks and then check if the failure is not in one of them like we do in the running status checks feedback and for that we need the API call to get all required checks. 

I think we missed this part when we introduced the concept of requiring only the required checks to pass.

Reply:
You mean `fetchAllStatusChecks` is undocumented API? (if so, worth a comment in code)

I thought the only undocumented API we were using was for the `commitStatus.mergeState: MergeState` that we use after the call to `fetchCommitStatus`+`fetchPullRequest`, not the `CommitState.State`? 🤔  So in that sense wouldn't getting rid of the undocumented API for `MergeState` and only relying on `CommitState.State` returned by `fetchAllStatusChecks` be better?

Reply:
> I wouldn't go there considering that what we are using is undocumented APIs

I don't think that's the case here @ilyapuchka, the only undocumented feature that we are using is the merge state part of the pull request, the API calls and combining the state of each status into a single one is something that we can do.

Reply:
I can look into that but personally I'd prefer to ship this fix faster as we have to keep answering people about this bug fixes that this change and they are affected by it quite ofter, and do optimisations later. I don't see how getting rid of one API call will significantly improve anything here.

Reply:
We are not talking about getting rid of one API call Ilya but something that's currently lacking to honour the configuration that we have to only check the required checks instead of all of them but we can address that in a separate pull request.

Reply:
In any case, tests are failing so we cannot ship the fix yet.

Reply:
> We are not talking about getting rid of one API call Ilya but something that's currently lacking to honour the configuration that we have to only check the required checks instead of all of them but we can address that in a separate pull request.

FWIW [commit 1b486b0](https://github.com/babylonhealth/Wall-E/commit/1b486b0295c9825a14feef2c7388a418d1d49ee4#diff-39d0b4d1a105d5e23d5b7b2556fb51fdR409) in this PR did introduce the fetching of required status checks, not sure that logic was later removed in subsequent commit?

Reply:
yes, it was removed as it's not exactly what we want to fetch here.

Reply:
> it was removed as it's not exactly what we want to fetch here

How do handle the scenario where the configuration only requires required checks to pass and you have one non-required failing? `fetchCommitStatus` will return `failure` and consequently mark the integration has failed even tho users are ok in having optional checks failing.

Reply:
We now only check if any statuses are still pending before checking for commit status as we used to. If that logic is not sufficient then I think it falls into what you mentioned before that we oversaw with requires status checks configuration. My understanding was that this can be tackled separately as it's more than just what this fix addresses (a race condition between triggering checks and checking them), or did I misunderstood?

Reply:
But if there's a commit with that logic already in place why did we remove it? 😕 We will need to bring those changes anyway either in this PR or in a following one.

Reply:
As per our discussion in Slack I realised it was not exactly what we needed here, so I changed from getting required checks to getting all checks and only checking if anything is still pending
https://babylonhealth.slack.com/archives/GQBDJ9422/p1575374833013900

Reply:
Ok let's tackle this in a separate ticket.